### ISSUE TITLE
`{{date::Y}}` is cached as a fragment with the full request URL as context

### ISSUE THREAD

**Issue by MartinUlisses** [contao/contao#7178]
### Affected version(s)

4.13.35 4.13.43

### Description

We use `{{date::Y}}` in the footer of our pages, as many other sites too. This week I became aware of a really high volume of cache data (20GB+ within the last 3 month), most of them expiring exactly at the end of the year. But caching was not enabled for any page.

What I realized is that most of the cache files have their origin in the `{{date::Y}}` insert tag. Documentation states that `{{date}}` is not cached. But the source code contains an exception (https://github.com/contao/contao/blob/4.13.43/core-bundle/src/Controller/InsertTagsController.php#L48). I assume the exception was made to limit the expiration date of the rendered pages using the tag.

The problem with the current implementation seems to be that the `{{date::Y}}` is cached with the full request URL as context. Thus it's cached once per page.

The really problematic thing is that the URL can contain parameters (as our search page does), which adds multiple cache files per set of parameters. (see cache file below)

```
1735685999
md6170e1b3c960785138f132655696dec29b1d569dfd5f77733921f791c1ba023d
a:1:{
	s:64:"a23a932a9b7b4872a07e77195216d504f9562582dbf6994996ef152644073438";a:4:{
		s:4:"vary";a:1:{i:0;s:6:"Origin";}
		s:7:"headers";a:12:{
			s:13:"cache-control";a:1:{i:0;s:6:"public";}
			s:4:"date";a:1:{i:0;s:29:"Tue, 30 Apr 2024 02:45:55 GMT";}
			s:7:"expires";a:1:{i:0;s:29:"Tue, 31 Dec 2024 22:59:59 GMT";}
			s:22:"x-content-type-options";a:1:{i:0;s:7:"nosniff";}
			s:15:"referrer-policy";a:1:{i:0;s:59:"no-referrer-when-downgrade, strict-origin-when-cross-origin";}
			s:18:"permissions-policy";a:1:{i:0;s:18:"interest-cohort=()";}
			s:12:"content-type";a:1:{i:0;s:24:"text/html; charset=UTF-8";}
			s:15:"x-frame-options";a:1:{i:0;s:10:"SAMEORIGIN";}
			s:16:"x-xss-protection";a:1:{i:0;s:13:"1; mode=block";}
			s:4:"vary";a:1:{i:0;s:6:"Origin";}
			s:16:"x-content-digest";a:1:{i:0;s:66:"en6557739a67283a8de383fc5c0997fbec7c5721a46f28f3235fc9607598d9016b";}
			s:14:"content-length";a:1:{i:0;i:4;}}
			s:6:"status";i:200;
			s:3:"uri";s:404:"https://****************/_fragment?_hash=43bJ%2BcHb75DQsgvhbnAHRUaQrDd5i4lc%2FPp5KC7j%2FfI%3D&_path=insertTag%3D%257B%257Bdate%253A%253AY%257D%257D%26_scope%3Dfrontend%26_format%3Dhtml%26_locale%3Dde%26_controller%3DContao%255CCoreBundle%255CController%255CInsertTagsController%253A%253ArenderAction&clientCache=0&pageId=10331&request=suche.html%3Fkeywords%3DDere%26page_s25%3D41%26page_s70%3D190";
	}
}
```

And that is what happens on our site. Most of the calls are made from "amazon-searchbot" and other bots calling our search page many many times per day with alternating search parameters.

I can work around the problem by modifying the page and use `{{date::Y|uncached}}` or `{{fragment::{{date::Y}}}}`, but this is not intuitive and can be easily forgotten, nor does the documentation contain any hint that `{{date::Y}}` is cached.

I also do not see any use in caching it once per page, as there is no relationship that would influence the outcome of `{{date::Y}}`. I would even think that not caching it at all and just using the expiration date would have a better performance (if possible).

What is the intended behaviour? I can't really tell.

**Comment by fritzmg** [comment]
The behaviour is intentional. `{{date::Y}}` is generated as a cacheable fragment, so that its cache can expire at the end of the year without having to invalidating the cache for the whole page.

However, fragments always need to be cached per URL so I am not sure if the high cache volume caused by this can be easily fixed while retaining this behaviour.

**Comment by MartinUlisses** [comment]
I understand that it needs to be cached for "a" URL, but does this URL really have to be different? At least that is how I understood the ESI mechanism.

Anyways, thanks for looking into it.

**Comment by fritzmg** [comment]
> I understand that it needs to be cached for "a" URL, but does this URL really have to be different? At least that is how I understood the ESI mechanism.

A fragment's output can be different for every URL. But may be this can be optimized.

**Comment by MartinUlisses** [comment]
> A fragment's output can be different for every URL.

A fragment's output in general surely can differ, but not for `{{date::Y}}`, as it renders the year part of "now".

Just my two cents. ;)

**Comment by fritzmg** [comment]
Yes, but in order take advantage of caching `{{date::Y}}` is rendered as a cacheable fragment, so that at the end of the year this part of the page gets updated correctly, without affecting the cache of the page itself.

**Comment by MartinUlisses** [comment]
I'm not sure how the cache proxy deals with it, but if this would mean that the cache proxy has to reassemble the page on every call, this cannot be cheaper than having the page to expire a the end of the year and then recreate it.

But I'm not exactly sure how that all works in detail and what is possible and what not. I just try to understand it.

**Comment by fritzmg** [comment]
That's how ESI fragments work. The proxy will make a request for every ESI fragment - but since the fragment is cached it will be fast.

**Comment by fritzmg** [comment]
> Most of the calls are made from "amazon-searchbot" and other bots calling our search page many many times per day with alternating search parameters.

Btw. you should set the search result page to `noindex` and also adjust your `robots.txt` in the website root accordingly, to alleviate this issue.

**Comment by leofeyer** [comment]
As discussed in the Contao call, this could be optimized in Contao 5.3 because we can add headers to insert tags there. An easy fix for Contao 4.13 would be not to cache the insert tag if the page cache is not active.

**Comment by MartinUlisses** [comment]
> Btw. you should set the search result page to `noindex` and also adjust your `robots.txt` in the website root accordingly, to alleviate this issue.

Thanks for the advice.

**Comment by MartinUlisses** [comment]
> As discussed in the Contao call, this could be optimized in Contao 5.3 because we can add headers to insert tags there. An easy fix for Contao 4.13 would be not to cache the insert tag if the page cache is not active.

Thanks for the quick response. I already added `|uncached` as a workaround.

**Comment by leofeyer** [comment]
As further discussed, we want to replace the following line with `$tag == 'date::Y' ? 'inline' : 'esi',`:

https://github.com/contao/contao/blob/a76ba264c6a51aff9064164f8993c874433dc536/core-bundle/src/Resources/contao/library/Contao/InsertTags.php#L230

We also have to add `$response->headers->set(SubrequestCacheSubscriber::MERGE_CACHE_HEADER, '1');` here:

https://github.com/contao/contao/blob/a76ba264c6a51aff9064164f8993c874433dc536/core-bundle/src/Controller/InsertTagsController.php#L49-L51
